plugins/gitlab/server/gitlab.ts

```diff
@@ -0,0 +1,430 @@
+import fetch, { RequestInit } from "node-fetch";
+import {
+  IntegrationService,
+  IntegrationType,
+  UnfurlResourceType,
+} from "@shared/types";
+import Logger from "@server/logging/Logger";
+import { Integration, IntegrationAuthentication, User } from "@server/models";
+import { UnfurlIssueOrPR, UnfurlSignature } from "@server/types";
+import { GitLabUtils } from "../shared/GitLabUtils";
+import env from "./env";
+import z from "zod";
+
+type MR = {
+  iid: number;
+  title: string;
+  description: string;
+  web_url: string;
+  state: string;
+  draft: boolean;
+  merged_at: string | null;
+  created_at: string;
+  author: {
+    username: string;
+    avatar_url: string;
+  };
+  labels: string[];
+};
+
+type Issue = {
+  iid: number;
+  title: string;
+  description: string;
+  web_url: string;
+  state: string;
+  created_at: string;
+  author: {
+    username: string;
+    avatar_url: string;
+  };
+  labels: { name: string; color: string }[];
+};
+
+const AccessTokenResponseSchema = z.object({
+  access_token: z.string(),
+  token_type: z.string(),
+  expires_in: z.number(),
+  refresh_token: z.string(),
+  scope: z.string(),
+  created_at: z.number(),
+});
+
+const UserInfoResponseSchema = z.object({
+  id: z.number(),
+  username: z.string(),
+  name: z.string(),
+  avatar_url: z.string().url(),
+});
+
+const projectsSchema = z.array(
+  z.object({
+    id: z.number(),
+    name: z.string(),
+    namespace: z.object({
+      id: z.number(),
+      full_path: z.string(),
+    }),
+  })
+);
+
+const ApplicationSchema = z.object({
+  id: z.number(),
+  application_id: z.string(),
+  application_name: z.string(),
+  callback_url: z.string().url(),
+  confidential: z.boolean(),
+});
+
+export class GitLab {
+  private static clientId = env.GITLAB_CLIENT_ID;
+  private static clientSecret = env.GITLAB_CLIENT_SECRET;
+
+  private static supportedResources = [
+    UnfurlResourceType.Issue,
+    UnfurlResourceType.PR,
+  ];
+
+  /**
+   * Parses a given URL and returns resource identifiers for GitLab specific URLs
+   *
+   * @param url URL to parse
+   * @returns Containing resource identifiers - `owner`, `repo`, `type` and `id`.
+   */
+  public static parseUrl(url: string) {
+    const { hostname, pathname } = new URL(url);
+    if (hostname !== "gitlab.com") {
+      return;
+    }
+
+    // GitLab URLs: /owner/repo/-/issues/123 or /owner/repo/-/merge_requests/123
+    const parts = pathname.split("/").filter(Boolean);
+    if (parts.length < 4) {
+      return;
+    }
+
+    const owner = parts[0];
```

Comment:
This is not reliably the owner, e.g. in my case it's a group:

<img width="681" height="601" alt="image" src="https://github.com/user-attachments/assets/725db73b-59b7-4ddc-91bd-88721772f7ef" />